arangod/Aql/SortExecutor.cpp

```diff
@@ -135,146 +134,129 @@ SortExecutor::~SortExecutor() {
-  TRI_ASSERT(_input == nullptr);
-  ExecutorState state;
-
-  // We need to get data
-  std::tie(state, _input) = inputMatrix.getMatrix();
+void SortExecutor::consumeInput(AqlItemBlockInputRange& inputRange,
+                                ExecutorState& state) {
+  size_t memoryUsageForRowIndexes =
+      inputRange.countDataRows() * sizeof(AqlItemMatrix::RowIndex);
 
-  // If the execution state was not waiting it is guaranteed that we get a
-  // matrix. Maybe empty still
-  TRI_ASSERT(_input != nullptr);
-  if (_input == nullptr) {
-    THROW_ARANGO_EXCEPTION(TRI_ERROR_OUT_OF_MEMORY);
-  }
-  // After allRows the dependency has to be done
-  TRI_ASSERT(state == ExecutorState::DONE);
+  ResourceUsageScope guard(_infos.getResourceMonitor(),
+                           memoryUsageForRowIndexes);
 
-  // Execute the sort
-  doSorting();
+  InputAqlItemRow input{CreateInvalidInputRowHint{}};
 
-  // If we get here we have an input matrix
-  // And we have a list of sorted indexes.
-  TRI_ASSERT(_input != nullptr);
-  TRI_ASSERT(_sortedIndexes.size() == _input->size());
-};
+  while (inputRange.hasDataRow()) {
+    // This executor is passthrough. it has enough place to write.
+    _rowIndexes.emplace_back(
+        std::make_pair(static_cast<std::uint32_t>(_inputBlocks.size() - 1),
+                       static_cast<std::uint32_t>(inputRange.getRowIndex())));
+    std::tie(state, input) =
+        inputRange.nextDataRow(AqlItemBlockInputRange::HasDataRow{});
+    TRI_ASSERT(input.isInitialized());
+  }
+  guard.steal();
+  _memoryUsageForRowIndexes += memoryUsageForRowIndexes;
+}
 
 std::tuple<ExecutorState, NoStats, AqlCall> SortExecutor::produceRows(
-    AqlItemBlockInputMatrix& inputMatrix, OutputAqlItemRow& output) {
+    AqlItemBlockInputRange& inputRange, OutputAqlItemRow& output) {
   AqlCall upstreamCall{};
-
-  if (!inputMatrix.hasDataRow()) {
-    // If our inputMatrix does not contain all upstream rows
-    return {inputMatrix.upstreamState(), NoStats{}, upstreamCall};
-  }
-
-  if (_input == nullptr) {
-    initializeInputMatrix(inputMatrix);
+  if (!_inputReady) {
+    auto inputBlock = inputRange.getBlock();
+    if (inputBlock != nullptr) {
+      _inputBlocks.emplace_back(inputBlock);
+    }
   }
 
-  if (_returnNext >= _sortedIndexes.size()) {
+  if (_returnNext >= _rowIndexes.size() && !_rowIndexes.empty()) {
     // Bail out if called too often,
     // Bail out on no elements
     return {ExecutorState::DONE, NoStats{}, upstreamCall};
   }
 
-  while (_returnNext < _sortedIndexes.size() && !output.isFull()) {
-    InputAqlItemRow inRow = _input->getRow(_sortedIndexes[_returnNext]);
+  ExecutorState state = ExecutorState::HASMORE;
+
+  if (!_inputReady) {
+    consumeInput(inputRange, state);
+    if (inputRange.upstreamState() == ExecutorState::HASMORE) {
+      return {state, NoStats{}, upstreamCall};
+    }
+    if (_returnNext < _rowIndexes.size()) {
```

Comment:
Is this `if` condition still needed? When we get here, `_returnNext` should always be `0`.

Reply:
Addressed